bootcompiler/src/main/scala/org/mozartoz/bootcompiler/transform/VariableClearing.scala

```diff
@@ -0,0 +1,312 @@
+        BranchClearsMap(this.clearsMap, ClearSite(falseStat, true))
+      })
+      walkExpr(condition)
+      
+    case MatchStatement(value, clauses, elseStatement) =>
+      /* If walkExpr was used, the variable would be cleared in-place, which is not desirable because
+       * it will be copied in the final AST, and so might clear itself before being read again */
+      def clearVal(stat: Statement) = {
+        val sym = value.asInstanceOf[Variable].symbol
+        if (this.clearsMap.get(sym) == Some(null))
+          this.clearsMap.put(sym, Seq(ClearSite(stat, true)))
+      }
+      
+      merge((clauses.map { clause =>
+        withClearsMap(undecided) {
+        	walkStat(clause.body)
+        	clause.guard.foreach(walkExpr(_))
+        	walkExpr(clause.pattern)
+        	clearVal(clause.body)
+        	BranchClearsMap(this.clearsMap, ClearSite(clause.body, true))
+        }
+      } :+ withClearsMap(undecided) {
+        walkStat(elseStatement)
+        clearVal(elseStatement)
+        BranchClearsMap(this.clearsMap, ClearSite(elseStatement, true))
+      }): _*)
+      
+    case _ => super.walkStat(statement)
+  }
+  
+  override def walkExpr(expression: Expression) = expression match {
+    case v @ Variable(sym) =>
+      if (clearsMap.get(sym) == Some(null)) {
+        clearsMap.put(sym, Seq(ClearSite(v, before=false)))
+      }
+      
+    case LocalExpression(decls, expr) =>
+      decls.foreach { decl =>
+        clearsMap.put(decl.symbol, null)
+      }
+      walkExpr(expr)
+      
+    case ProcExpression(name, args, body, flags) =>
+      val walker = new CaptureWalker(this.clearsMap)
+      walker.walkStat(body)
+      val captured = walker.captured
+      for (sym <- walker.captured) {
+        clearsMap.put(sym, Seq(ClearSite(expression, before=false)))
+      }
+      
+    case IfExpression(condition, trueExpr, falseExpr) =>
+      merge(withClearsMap(undecided) {
+        walkExpr(trueExpr)
+        BranchClearsMap(this.clearsMap, ClearSite(trueExpr, true))
+      }, withClearsMap(undecided) {
+    	  walkExpr(falseExpr)
+        BranchClearsMap(this.clearsMap, ClearSite(falseExpr, true))
+      })
+      walkExpr(condition)
+      
+    case ShortCircuitBinaryOp(left, "andthen", right) => // We use the substitution mechanism, only if necessary.
+      val subst = treeCopy.IfExpression(expression, left, right, Constant(False())(right.pos))
+      walkExpr(subst)
+      nodeSubst.put(expression, subst)
+      
+    case ShortCircuitBinaryOp(left, "orelse", right) =>
+      val subst = treeCopy.IfExpression(expression, left, Constant(True())(right.pos), right)
+      walkExpr(subst)
+      nodeSubst.put(expression, subst)
+      
+    case MatchExpression(value, clauses, elseExpression) =>
+      def clearVal(expr: Expression) = {
+        val sym = value.asInstanceOf[Variable].symbol
+        if (this.clearsMap.get(sym) == Some(null))
+          this.clearsMap.put(sym, Seq(ClearSite(expr, true)))
+      }
```

Comment:
=> the Translator should create its own variables to avoid leaking details here.